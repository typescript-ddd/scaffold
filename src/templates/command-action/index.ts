export * from "./command-action.generator";
export * from "./command-action.template";
export * from "./command-action.types";