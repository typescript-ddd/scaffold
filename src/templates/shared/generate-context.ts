import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}