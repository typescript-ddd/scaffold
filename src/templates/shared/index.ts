export * from "./context-builder";
export * from "./generate-context";