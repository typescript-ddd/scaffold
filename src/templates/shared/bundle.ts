import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
