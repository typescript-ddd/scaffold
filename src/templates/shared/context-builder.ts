import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
