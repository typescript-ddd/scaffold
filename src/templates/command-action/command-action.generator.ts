import { Project } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { CommandActionTemplateValues } from "./command-action.types";

export const generateCommandAction = (
  values: Partial<CommandActionTemplateValues>,
  context: GenerateContext
) => {
  const { subject, method, path, requestType, responseType, contextType } =
    values;
  const methodType = strings.capitalize(strings.lower(method));
  const className = `${methodType}${strings.capitalize(subject)}Action`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} extends CommandAction {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["CommandAction", "CommandBus"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [{ name: "commandBus", type: "CommandBus", isReadonly: true }],
  });
  constructorDeclaration.setBodyText((writer) => {
    writer.writeLine(`super(commandBus);`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      className
    )}.`,
    tags: [
      { tagName: "param", text: `{CommandBus} commandBus - The command bus.` },
    ],
  });

  classDeclaration.addProperty({
    name: "method",
    type: "string",
    initializer: `"${method}"`,
    isReadonly: true,
  });

  classDeclaration.addProperty({
    name: "path",
    type: "string",
    initializer: `"${path}"`,
    isReadonly: true,
  });

  const executeDeclaration = classDeclaration.addMethod({
    name: "execute",
    parameters: [
      { name: "req", type: requestType || "any" },
      { name: "res", type: responseType || "any" },
      { name: "ctx", type: contextType || "any" },
    ],
    returnType: "Promise<void>",
    isAsync: true,
  });

  executeDeclaration.setBodyText((writer) => {
    writer.writeLine(`// TODO: Implement action logic.`);
    writer.writeLine(`res.json({ message: "Hello, World!" });`);
  });

  executeDeclaration.addJsDoc({
    description: `Executes the ${strings.describe(className)}.`,
    tags: [
      {
        tagName: "param",
        text: `{${requestType || "any"}} req - The request object.`,
      },
      {
        tagName: "param",
        text: `{${responseType || "any"}} res - The response object.`,
      },
      {
        tagName: "param",
        text: `{${contextType || "any"}} ctx - The context object.`,
      },
    ],
  });

  let output = sourceFile.getFullText();

  return output;
};
