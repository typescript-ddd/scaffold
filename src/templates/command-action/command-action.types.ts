import { Maybe, TemplateStringOption } from "../shared/template";

export type CommandActionTemplateValues = {
  subject: string;
  method: string;
  path: string;
  requestType: Maybe<string>;
  responseType: Maybe<string>;
  contextType: Maybe<string>;
};

export type CommandActionTemplateOptions = {
  subject: TemplateStringOption;
  method: TemplateStringOption;
  path: TemplateStringOption;
  requestType: TemplateStringOption;
  responseType: TemplateStringOption;
  contextType: TemplateStringOption;
};
