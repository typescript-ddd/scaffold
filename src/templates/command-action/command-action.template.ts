import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateCommandAction } from "./command-action.generator";
import {
  CommandActionTemplateOptions,
  CommandActionTemplateValues,
} from "./command-action.types";

export class CommandActionTemplate
  implements
    Template<CommandActionTemplateValues, CommandActionTemplateOptions>
{
  readonly name = "Command Action";
  readonly description = "Generates a command action";
  readonly options: CommandActionTemplateOptions = {
    subject: {
      name: "subject",
      label: "Subject",
      description: "The subject of the command action",
      optionType: "string",
    },
    method: {
      name: "method",
      label: "Method",
      description: "The method of the command action",
      optionType: "string",
    },
    path: {
      name: "path",
      label: "Path",
      description: "The path of the event",
      optionType: "string",
    },
    contextType: {
      name: "contextType",
      label: "Context Type",
      description: "The type of the context",
      optionType: "string",
    },
    requestType: {
      name: "requestType",
      label: "Request Type",
      description: "The type of the request",
      optionType: "string",
    },
    responseType: {
      name: "responseType",
      label: "Response Type",
      description: "The type of the response",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    subject: "User",
    method: "create",
    path: "UserCreated",
    requestType: "CreateUserRequest",
    responseType: "CreateUserResponse",
    contextType: "UserContext",
  };
  generate(values: CommandActionTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateCommandAction(values, context, chunkName);
  }
}
