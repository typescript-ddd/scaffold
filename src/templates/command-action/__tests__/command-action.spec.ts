import { ContextBuilder } from "../../shared";
import { generateCommandAction } from "../command-action";

describe("CommandAction", () => {
    it("should generate command action", () => {
        // Arrange
        const options = {
            method: "GET",
            path: "/user/:id",
            subject: "User",
        };
        const context = ContextBuilder.build("@src");

        // Act
        const output = generateCommandAction(options, context);

        // Assert
        expect(output).toMatchSnapshot();
    }); 
})