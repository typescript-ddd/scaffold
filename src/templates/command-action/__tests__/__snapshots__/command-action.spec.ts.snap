// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandAction should generate command action 1`] = `
"import { CommandAction, CommandBus } from "@typescript-ddd/core";

/** Represents a get user action. */
export class GetUserAction extends CommandAction {
    /**
     * Initializes a new instance of a get user action.
     * @param {CommandBus} commandBus - The command bus.
     */
    constructor(readonly commandBus: CommandBus) {
        super(commandBus);
    }

    readonly method: string = "GET";
    readonly path: string = "/user/:id";

    /**
     * Executes the a get user action.
     * @param {any} req - The request object.
     * @param {any} res - The response object.
     * @param {any} ctx - The context object.
     */
    async execute(req: any, res: any, ctx: any): Promise<void> {
        // TODO: Implement action logic.
        res.json({ message: "Hello, World!" });
    }
}    
"
`;
